src/kudu/consensus/consensus_meta.cc

```diff
@@ -173,9 +173,23 @@ const string& ConsensusMetadata::leader_uuid() const {
   return leader_uuid_;
 }
 
+const LastKnownLeader& ConsensusMetadata::last_known_leader() const {
+  DFAKE_SCOPED_RECURSIVE_LOCK(fake_lock_);
+  return last_known_leader_;
+}
+
 void ConsensusMetadata::set_leader_uuid(string uuid) {
   DFAKE_SCOPED_RECURSIVE_LOCK(fake_lock_);
   leader_uuid_ = std::move(uuid);
+
+  // Only update last_known_leader_ when the current node
+  // 1) has won a leader election (LEADER)
+  // 2) receives AppendEntries from a legitimate leader (FOLLOWER)
+  if (!uuid.empty()) {
```

Comment:
didn't you just move uuid to leader_uuid_ ?

Reply:
Kudu uses move semantics because it works better with inlining optimizations. Clang tidy and other linters, in addition to UBSAN, will catch these types of bugs so it's only a risk to us because we don't have CI. Anyway good catch.

Reply:
> I am curious about this but I will follow up a little later. If there is an article somewhere, I would love to read it.

Here's one: https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html